File: Grind 75 Questions/Grind 75 using Python/2. Valid Parentheses/Sol 1.py
# Time Complexity: O(n)

class Solution:
    def isValid(self, s: str) -> bool:
        close = {
            ')': '(',
            '}': '{',
            ']': '['}
        p_stack = list()
        for pr in s:
            if pr in close:
                if len(p_stack) != 0 and p_stack[-1] == close[pr]:
                    p_stack.pop()
                else:
                    return False
            else:
                p_stack.append(pr)
        return True if len(p_stack) == 0 else False

File: Blind-75-Questions/Java/55.Valid-Parentheses/Sol-1.java
class Solution {
    public static boolean arePairs(char open, char close){
        if (open == '(' && close == ')')
            return true;
        else if (open == '[' && close == ']')
            return true;
        else return open == '{' && close == '}';
    }

    public static boolean areBalanced(String exp) {
        Stack<Character> stack = new Stack<>();
        if (exp.length() % 2.0 != 0) return false;
        for (char c : exp.toCharArray()) {
            if (c == '(' || c == '[' || c == '{')
                stack.push(c);
            else if (!stack.isEmpty() && arePairs(stack.peek(), c)) {
                stack.pop();
            }
            else return false;
        }
        return stack.isEmpty();
    }
    
    public boolean isValid(String s) {
        return areBalanced(s);
    }
}